src/mbi/relaxed_projection.py

```diff
@@ -0,0 +1,368 @@
+import numpy as np
+import pandas as pd
+import itertools
+from typing import Callable, List, Tuple
+
+import jax
+import jax.numpy as jnp
+import optax
+from .dataset import Dataset
+from .domain import Domain
+
+
+
+#######################################################
+#                                                     #
+#       Here is a the projectable data class          #
+#                                                     #
+#######################################################
+
+class ProjectableData():
+    def __init__(
+            self, 
+            D: jax.Array, 
+            domain: Domain
+        ):
+        """
+        A class storing one-hot data, supporting projection, vector 
+        query, and synthesize data operation.
+        
+        Args:
+            D: a jax.Array, representing a one-hot form of the data.
+            domain: a Domain class.
+        """
+        self.D_prime = D
+        self.domain = domain
+        self.feature_indices_map = {}
+        s = 0
+        for col in self.domain.attrs:
+            e = s + self.domain.size(col)
+            self.feature_indices_map[col] = list(range(s, e))
+            s = e
+
+    def datavector(self):
+        splits = [self.D_prime[:, self.feature_indices_map[attr]] for attr in self.domain.attrs]
+
+        letters = "bcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+        assert(
+            len(letters) >= len(self.domain)
+        ), 'High dim domain'
+
+        input_dims = ','.join(f'a{letters[i]}' for i in range(len(splits)))
+        output_dims = ''.join(f'{letters[i]}' for i in range(len(splits)))
+        einsum_str = f'{input_dims}->a{output_dims}'
+
+        joint = jnp.einsum(einsum_str, *splits).mean(axis=0)
+        return np.array(joint).flatten()
```

Comment:
what's the reason for converting to numpy here?  I'd probably favor keeping this as jax arrays unless there's a compelling reason not to (in which case you should add an inline comment explaining the decision)